Add support for `--before`, `--then` and `--else` flags

FungusHumungus opened the pull request:
Fix #410

### Description of the change

Added support for commands to be run around the build. `--before` is run before the build, `--then` following a successful build and `--else` following an unsuccessful build.

I realise there were some alternative suggestions in the issue, but I personally prefer the original suggestion as I really liked the way it worked with Pulp.

### Checklist:

- [*] Added the change to the "Unreleased" section of the changelog
- [*] Added some example of the new feature to the `README`
- [*] Added a test for the contribution (if applicable)

**P.S.**: the above checks are not compulsory to get a change merged, so you may skip them. However, taking care of them will result in less work for the maintainers and will be much appreciated 😊

f-f commented on README.md:
````diff
@@ -303,6 +303,25 @@ $ spago build --watch
+To run a command before a build you can use the `--before` flag, eg to clear the screen before a build:
+
+```bash
+$ spago build --watch --before clear
````
To clear the screen we already have the `--clear-screen` option, could we pick an alternative example here? Or alternatively show them side-by-side

FungusHumungus replied:
Good point. I'll try to think of an alternative example.

FungusHumungus commented:
It has also occurred to me that the build should be cancelled if the command fails in some way. Currently it is ignoring the result.

FungusHumungus commented:
I have changed the example just just post a notification. Inspiration for anything else just wasn't coming!

Now if the command returns an error, it will halt the build.

I have also moved the `then` command so it runs after the `postBuild`. I feel it makes more sense to run this after bundling has completed when running `bundle-app`. It could be useful to do some processing with the bundled file.

Let me know if you want me to revert either of these additional changes.

f-f commented:
@FungusHumungus all the new changes look great, thank you for doing this! :slightly_smiling_face:

FungusHumungus commented:
It was a pleasure! :-)
